feat: auto update approval (#869)
sse-secure-systems/connaisseur: docs/features/automatic_update_approval.md

# Automatic Update Approval

> With the feature *automatic update approval* enabled, Connaisseur now may automatically approve any resource that is updated and doesn't change its image references. This is especially useful when handling long living resources, with potentially out of sync signature data, that still need to be scaled up and down.

Reviewer: ```suggestion
With the feature *automatic update approval* enabled, Connaisseur automatically approves any resource that is updated and doesn't change its image references. This is especially useful when handling long living resources, with potentially out-of-sync signature data, that still need to be scaled up and down.
```

> With the feature *automatic update approval* enabled, Connaisseur now may automatically approve any resource that is updated and doesn't change its image references. This is especially useful when handling long living resources, with potentially out of sync signature data, that still need to be scaled up and down.

Reviewer: Think we had policy of newline after each sentence for better updates (?) 🤷‍♂️

> An example: When dealing with a deployment that has an image reference `image:tag`, this reference is updated by Connaisseur during signature validation to `image@sha256:123...`, to ensure the correct and secure image is used by the deployment. When scaling up or down the deployment, the image reference `image@sha256:123...` is presented to Connaisseur, due to the updated. Now over the time the signature of the original `image:tag` may change and a new "secure" image is available at `image@sha256:456...`. If at that time the deployment in scaled up or down, Connaisseur will try to validate the image reference `image@sha256:123...`, by looking for it inside the signature data it receives. Unfortunately this reference may no longer be present, due to signature updates and thus the whole up or down scaling will be denied.

Reviewer: ```suggestion
An example: When dealing with a deployment that has an image reference `image:tag`, this reference is updated by Connaisseur during signature validation to `image@sha256:123...`, to ensure the correct image is used by the deployment. When scaling up or down the deployment, the image reference `image@sha256:123...` is presented to Connaisseur, due to the updated definition. Over time the signature of the original `image:tag` may change and a new "correct" image is available at `image@sha256:456...`. If afterwards the deployment in scaled up or down, Connaisseur will try to validate the image reference `image@sha256:123...`, by looking for it inside the signature data it receives. Unfortunately this reference may no longer be present, due to signature updates and thus the whole scaling will be denied.
```

> With automatic update approval, this is no longer the case. The validation of `image@sha256:123...` will be skipped, as only the number of pods is changed and no different image is used.

Reviewer: ```suggestion
With automatic update approval enabled, this is no longer the case. The validation of `image@sha256:123...` will be skipped, as no different image is used.
```

> This obviously has security implications, since it's no longer guaranteed that resources that are updated, have fresh and up-to-date signatures. So use it with caution.

Reviewer: Would add we default to disable the feature

> This obviously has security implications, since it's no longer guaranteed that resources that are updated, have fresh and up-to-date signatures. So use it with caution.

Reviewer: ```suggestion
This obviously has security implications, since it's no longer guaranteed that resources that are updated, have fresh and up-to-date signatures. So use it with caution. The creation of resources on the other hand remains unchanged and will enforce validation.
```
or sth like that?

## Configuration options

`automaticUpdateApproval` in `helm/value.yaml` supports the following keys:

| Key | Default | Required | Description |
| - | - | - | - |
| `automaticUpdateApproval` | false |  | `true` or `false` ; when `true`, Connaisseur will enable automatic update approval |

## Example

In `helm/values.yaml`:

```yaml

> automaticUpdateApproval: true

Reviewer: I think for v3, we should change these to be below a common thingy

admissionStrategies:
 - autoChild
 - autoUpdate

Author: i thought about a unified `feature` field in which `detectionMode`, `namespacedValidation` and `autoChild` `autoUpdate` reside.

Reviewer: Or sth like that 👍 
out of scope here anyway, but good to know you're feeling the same

Over time the signature of the original `image:tag` may change and a new "correct" image is available at `image@sha256:456...`.
If afterwards the deployment in scaled up or down, Connaisseur will try to validate the image reference `image@sha256:123...`, by looking for it inside the signature data it receives.
Unfortunately this reference may no longer be present, due to signature updates and thus the whole scaling will be denied.

With automatic unchanged approval enabled, this is no longer the case.
The validation of `image@sha256:123...` will be skipped, as no different image is used.

This obviously has security implications, since it's no longer guaranteed that resources that are updated, have fresh and up-to-date signatures.
So use it with caution.
For that reason the feature is also disabled by default.
The creation of resources on the other hand remains unchanged and will enforce validation.

## Configuration options

`automaticUnchangedApproval` in `helm/value.yaml` supports the following keys:

| Key | Default | Required | Description |
| - | - | - | - |

> | `automaticUnchangedApproval` | false |  | `true` or `false` ; when `true`, Connaisseur will enable automatic update approval |

Reviewer: still update approval
